blockstore/badger/blockstore.go

```diff
@@ -165,26 +183,342 @@ func (b *Blockstore) isOpen() bool {
+		return fmt.Errorf("error moving badger blockstore to %s: %w", path, err)
+	}
+
+	b.lockMove()
+	db1 := b.db
+	b.db = b.db2
+	b.db2 = nil
+	b.unlockMove(moveStateCleanup)
+
+	err = db1.Close()
+	if err != nil {
+		log.Warnf("error closing badger db: %s", err)
+	}
+
+	dbpath := b.opts.Dir
+	oldpath := fmt.Sprintf("%s.old.%d", dbpath, time.Now().Unix())
+
+	ok := true
+	err = os.Rename(dbpath, oldpath)
+	if err != nil {
+		// this is bad, but not catastrophic; new data will be written in db2 and user can fix
+		log.Errorf("error renaming badger db dir from %s to %s; USER ACTION REQUIRED", dbpath, oldpath)
+		ok = false
+	}
+
+	if ok {
+		err = os.Symlink(path, dbpath)
+		if err != nil {
+			// ditto, this is bad, but not catastrophic; user can fix
+			log.Errorf("error symlinking badger db dir from %s to %s; USER ACTION REQUIRED", path, dbpath)
+		}
+
+		b.deleteDB(oldpath)
+	}
+
+	log.Info("moving blockstore done")
+	return nil
+}
+
+// doCopy copies a badger blockstore to another, with an optional filter; if the filter
+// is not nil, then only cids that satisfy the filter will be copied.
+func (b *Blockstore) doCopy(from, to *badger.DB, filter func(cid.Cid) bool) error {
+	count := 0
+	batch := to.NewWriteBatch()
+	defer batch.Cancel()
+
+	txn := from.NewTransaction(false)
+	defer txn.Discard()
+
+	opts := badger.IteratorOptions{PrefetchSize: moveBatchSize}
+	iter := txn.NewIterator(opts)
+	defer iter.Close()
+
+	pooled := make([][]byte, 0, 2*moveBatchSize)
+	getPooled := func(size int) []byte {
+		buf := pool.Get(size)
+		pooled = append(pooled, buf)
+		return buf
+	}
+	putPooled := func() {
+		for _, buf := range pooled {
+			pool.Put(buf)
+		}
+		pooled = pooled[:0]
+	}
+	defer putPooled()
+
+	var buf []byte
+	for iter.Rewind(); iter.Valid(); iter.Next() {
+		if !b.isOpen() {
+			return ErrBlockstoreClosed
+		}
+
+		item := iter.Item()
+
+		kk := item.Key()
+		if filter != nil {
+			k := kk
+			if b.prefixing {
+				k = k[b.prefixLen:]
+			}
+
+			klen := base32.RawStdEncoding.DecodedLen(len(k))
+			if klen > len(buf) {
+				buf = make([]byte, klen)
+			}
+
+			n, err := base32.RawStdEncoding.Decode(buf, k)
+			if err != nil {
+				return err
+			}
+
+			c := cid.NewCidV1(cid.Raw, buf[:n])
+			if !filter(c) {
+				continue
+			}
+		}
+
+		k := getPooled(len(kk))
+		copy(k, kk)
+
+		var v []byte
+		err := item.Value(func(vv []byte) error {
+			v = getPooled(len(vv))
+			copy(v, vv)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		if err := batch.Set(k, v); err != nil {
+			return err
+		}
+
+		count++
+		if count == moveBatchSize {
+			if err := batch.Flush(); err != nil {
+				return err
+			}
+			// Flush discards the transaction, so we need a new batch
+			batch = to.NewWriteBatch()
+			count = 0
+			putPooled()
+		}
+	}
+
+	if count > 0 {
+		return batch.Flush()
+	}
+
+	return nil
+}
+
+func (b *Blockstore) deleteDB(path string) {
+	// follow symbolic links, otherwise the data wil be left behind
+	lpath := path
+	for {
```

Comment:
EvalSymlinks?

Reply:
aha! I didn't know this method existed; I'll use it.

Reply:
done.